### PR TITLE
Automate basic test scenario for TPP testing

### PR DESCRIPTION
**Pull request by Mahalty**
#### Describe Bug or Feature

-  Automate following scenario from TPP test suite. 
**Scenario:** Verify that the connection to the daemons is happening via TPP by matching the logs msgs in daemon logs . Submit jobs and a reservation in the tests to ensure end to end flow communication b/w server-mom and mom-mom through pbs_comm.

#### Attach Test and Valgrind Logs/Output

- [Execution_logs_TestWeanUdp.txt](https://github.com/PBSPro/pbspro/files/4354178/Execution_logs_TestWeanUdp.txt)

### CODE REVIEW

File: test/tests/functional/pbs_wean_udp.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+# disassemble the software. Usage of the software and other information
+# provided by Altair(or its resellers) is only as explicitly stated in the
+# applicable end user license agreement between Altair and Licensee.
+# In the absence of such agreement, the Altair standard end user
+# license agreement terms shall govern.
+
+from tests.functional import *
+import socket
+
+class TestWeanUdp(TestFunctional):
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: I would suggest the name `TestTPP` or `TestPbsComm`

File: test/tests/functional/pbs_wean_udp.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    Test suite consists of tests to check the functionality of pbs_comm daemon
+    """
+    def setUp(self):
+        TestFunctional.setUp(self)
+        #msg = "Need 2 mom hosts which are not present on server host"
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: Remove commented lines

File: test/tests/functional/pbs_wean_udp.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    def submit_job(self, set_attrib=None, exp_attrib=None, sleep=10, 
+                   interactive=False):
+        j = Job(PBSROOT_USER)
+        if set_attrib:
+           j.set_attributes(set_attrib)
+        if interactive:
+            j.interactive_script = [('hostname', '.*'),
+                                    ('export PATH=$PATH:%s' % self.exec_path, '.*'),
+                                     ('qstat', '.*')]
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: Please run autopep8 on file

File: test/tests/functional/pbs_wean_udp.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        return rid
+
+    def test_comm_with_mom(self):
+        """
+        Test the installation of communication daemon during PBS installation.
+        """
+        msg = "Need 2 moms as input. use -pmoms=<m1>:<m2>"
+        if len(self.moms) != 2:
+            self.skip_test(reason=msg)
+        (self.momA, self.momB) = self.moms.values()
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: this will fail if -pmoms is more than 2

File: test/tests/functional/pbs_wean_udp.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        self.hostA = self.momA.shortname
+        self.hostB = self.momB.shortname
+        log_msg = ["TPP initialization done",
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: `log_msgs`

File: test/tests/functional/pbs_wean_udp.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            self.server.log_match(msg, regexp=True)
+            self.scheduler.log_match(msg, regexp=True)
+            self.momA.log_match(msg, regexp=True)
+            self.momB.log_match(msg, regexp=True)
+        server_ip = socket.gethostbyname(self.server.hostname)
+        msg = "Registering address %s:15001 to pbs_comm" % server_ip
+        self.server.log_match(msg)
+        msg = "Registering address %s:15004 to pbs_comm" % server_ip
+        self.scheduler.log_match(msg)
+        for mom_name in self.moms.values():
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: `mom` instead `mom_name`

File: test/tests/functional/pbs_wean_udp.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        # Submit job into reservation
+        set_attrib = {ATTR_q: resv_que}
+        jid = self.submit_job(set_attrib, sleep=60)
+        # Wait for reservation to start
+        a = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|2')}
+        self.server.expect(RESV, a, rid, offset=5)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        self.server.delete(rid)
+        # wait for sometime for the job in reservation queue to get delete
+        time.sleep(5)
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_wean_udp.py]: I literally don't like this way of waiting for the job to vanish, what if the machine is too slow? will this time sufficient?
instead, we should use server.expect() with queue attr unset on job or check job end message in server/accounting log

**Comment by Mahalty** [comment]
**Latest execution-logs:**
[Execution_logs_TestTPP_3moms.txt](https://github.com/PBSPro/pbspro/files/4354373/Execution_logs_TestTPP_3moms.txt)

File: test/tests/functional/pbs_wean_udp.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    def submit_resv(self, set_attrib=None, exp_attrib=None):
+        r = Reservation(PBSROOT_USER)
+        if set_attrib:
+            r.set_attributes(set_attrib)
+        rid = self.server.submit(r)
+        if exp_attrib:
+            self.server.expect(RESV, exp_attrib, id=rid)
+        return rid
+
+    def test_comm_with_mom(self):
```
Comment (saritakh) [inline test/tests/functional/pbs_wean_udp.py]: Add @requirements decorator to this test that defines the specifics of number of each type of daemons needed for this test case.

File: test/tests/functional/pbs_test_tpp.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        if len(self.moms) < 2:
+            self.skip_test(reason=msg)
+        log_msgs = ["TPP initialization done",
+                    "Single pbs_comm configured, " +
+                    "TPP Fault tolerant mode disabled",
+                    "Connected to pbs_comm %s.*:17001" % self.server.shortname]
+        for msg in log_msgs:
+            self.server.log_match(msg, regexp=True)
+            self.scheduler.log_match(msg, regexp=True)
+        for mom in self.moms.values():
```
Comment (hirenvadalia) [inline test/tests/functional/pbs_test_tpp.py]: I think you want to put this for loop under above for loop?

File: test/tests/functional/pbs_wean_udp.py (hunk 4 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_wean_udp.py]: but in that case, we would have skipped the test, just before we could execute this.

**Comment by Mahalty** [comment]
Execution-logs:
[Execution_logs_TestTPP_3moms_1.txt](https://github.com/PBSPro/pbspro/files/4354530/Execution_logs_TestTPP_3moms_1.txt)

File: test/tests/functional/pbs_wean_udp.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            msg1 = "Registering address %s:15003 to pbs_comm" % ip
+            msg2 = "Leaf registered address %s:15003" % ip
+            mom_name.log_match(msg1)
+            self.comm.log_match(msg2)
+        # Submit job
+        jid = self.submit_job(exp_attrib={'job_state': 'R'})
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET, offset=10)
+        self.server.log_match("%s;Exit_status=0" % jid)
+        # Submit multi-chunk job
+        set_attrib = {'Resource_List.select': '2:ncpus=1'}
```
Comment (saritakh) [inline test/tests/functional/pbs_wean_udp.py]: I suggest to add 'scatter' placement to this job in order to convert it to multi node job with pbsdsh.

File: test/tests/functional/pbs_wean_udp.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        jid = self.submit_job(set_attrib, exp_attrib={'job_state': 'R'})
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET, offset=10)
+        self.server.log_match("%s;Exit_status=0" % jid)
+        # Submit interactive job
+        set_attrib = {ATTR_inter: ''}
+        jid = self.submit_job(set_attrib, interactive=True)
```
Comment (saritakh) [inline test/tests/functional/pbs_wean_udp.py]: suggest to add exp_attrib for interactive job or check its running state.

File: test/tests/functional/pbs_wean_udp.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        # Submit reservation
+        set_attrib = {'Resource_List.select': '1:ncpus=1',
```
Comment (saritakh) [inline test/tests/functional/pbs_wean_udp.py]: Suggest this reservation also to be multi chunk with 'scatter' placement.

File: test/tests/functional/pbs_test_tpp.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    """
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        self.pbs_conf = self.du.parse_pbs_config(self.server.shortname)
+        self.exec_path = os.path.join(self.pbs_conf['PBS_EXEC'], "bin")
+
+    def submit_job(self, set_attrib=None, exp_attrib=None, sleep=10,
+                   interactive=False):
+        j = Job(PBSROOT_USER)
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: why root user?

File: test/tests/functional/pbs_test_tpp.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        # Submit multi-chunk job
+        set_attrib = {'Resource_List.select': '2:ncpus=1'}
+        jid = self.submit_job(set_attrib, exp_attrib={'job_state': 'R'})
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET, offset=10)
+        self.server.log_match("%s;Exit_status=0" % jid)
+        # Submit interactive job
+        set_attrib = {ATTR_inter: ''}
+        jid = self.submit_job(set_attrib, interactive=True)
+        # Submit reservation
+        set_attrib = {'Resource_List.select': '1:ncpus=1',
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: do we need to provide select here?

File: test/tests/functional/pbs_test_tpp.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+                      'reserve_start': int(time.time() + 5),
+                      'reserve_end': int(time.time() + 120)}
+        exp_attrib = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        rid = self.submit_resv(set_attrib, exp_attrib)
+        resv_que = rid.split('.')[0]
+        # Submit job into reservation
+        set_attrib = {ATTR_q: resv_que}
+        jid = self.submit_job(set_attrib, sleep=60)
+        # Wait for reservation to start
+        a = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|2')}
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: it should be RESV_RUNNING|5

File: test/tests/functional/pbs_test_tpp.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        self.server.expect(RESV, a, rid, offset=5)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        self.server.delete(rid)
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET)
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: do we need the last 2 lines for testing TPP?

File: test/tests/functional/pbs_test_tpp.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        set_attrib = {'Resource_List.select': '2:ncpus=1'}
+        jid = self.submit_job(set_attrib, exp_attrib={'job_state': 'R'})
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET, offset=10)
+        self.server.log_match("%s;Exit_status=0" % jid)
+        # Submit interactive job
+        set_attrib = {ATTR_inter: ''}
+        jid = self.submit_job(set_attrib, interactive=True)
+        # Submit reservation
+        set_attrib = {'Resource_List.select': '1:ncpus=1',
+                      'reserve_start': int(time.time() + 5),
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: 5 seconds could be short if you want to check for the confirmed state. on slow machines the reservation may start running before you check for the confirmed state.

**Comment by Mahalty** [comment]
**Latest execution-logs:**
[Execution_logs_TestTPP_1.txt](https://github.com/PBSPro/pbspro/files/4358900/Execution_logs_TestTPP_1.txt)

File: test/tests/functional/pbs_test_tpp.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        if interactive:
+            j.interactive_script = [('hostname', '.*'),
+                                    ('export PATH=$PATH:%s' %
+                                     self.exec_path, '.*'),
+                                    ('qstat', '.*')]
+        j.set_sleep_time(sleep)
+        if job_script:
+            pbsdsh_path = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                                       "bin", "pbsdsh")
+            script = "#!/bin/sh\n%s sleep 10" % pbsdsh_path
```
Comment (prakashcv13) [inline test/tests/functional/pbs_test_tpp.py]: if job_script is true, sleep value passed is not used. I do not see a reason for them to be mutually exclusive.

**Comment by Mahalty** [comment]
**Latest execution-logs:**
[Execution_logs_TestTPP_2.txt](https://github.com/PBSPro/pbspro/files/4367948/Execution_logs_TestTPP_2.txt)

**Comment by Mahalty** [comment]
> Please address my comment given earlier -
> "please provide some detail on how the test is helping in testing TPP."

@prakashcv13: In the test we are verifying the msgs that are logged 

> Please address my comment given earlier -
> "please provide some detail on how the test is helping in testing TPP."

@prakashcv13 The test is initially verifying that the connection to the daemons is happening via TPP by matching the logs msgs in daemon logs . We are also submitting jobs in the tests to ensure end to end flow communication b/w server-mom and mom-mom through pbs_comm.

**Comment by prakashcv13** [comment]
@hirenvadalia, @saritakh - there have been changes after you approved.
Could you please check and let us know if the updated changes are good enough.

File: test/tests/functional/pbs_test_tpp.py [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        jid = self.submit_job(set_attrib, exp_attrib={'job_state': 'R'},
+                              job_script=True)
+        self.server.expect(JOB, 'queue', id=jid, op=UNSET, offset=10)
+        self.server.log_match("%s;Exit_status=0" % jid)
+        # Submit interactive job
+        set_attrib = {'Resource_List.select': '2:ncpus=1',
+                      ATTR_inter: '',
+                      ATTR_l + '.place': 'scatter'}
+        jid = self.submit_job(set_attrib, exp_attrib={'job_state': 'R'},
+                              interactive=True, job_script=True)
```
Comment (saritakh) [inline test/tests/functional/pbs_test_tpp.py]: Interactive job should not have job_script=True, right?

**Comment by saritakh** [comment]
@Mahalty the changes look good to me except the interactive job code, which has job_script=True.

**Comment by Mahalty** [comment]
**Latest execution-logs:**
[Execution_logs_TestTPP_4.txt](https://github.com/PBSPro/pbspro/files/4373348/Execution_logs_TestTPP_4.txt)

**Pull request merged**